### ISSUE TITLE
Warning spam when extensions are missing is excessive

### ISSUE THREAD

**Issue by gcp** [NVIDIA/apex#97]
Issue #96 does not concern me all too much, but during training I almost can't see the loss numbers in the repeated spam of things like

```
Warning:  apex was installed without --cuda_ext.  FusedAdam will be unavailable.
Warning:  apex was installed without --cuda_ext.  FusedLayerNorm will be unavailable.
```

I'm not sure how useful this warning is. But for sure, giving it more than once serves no purpose.

**Comment by mcarilli** [comment]
Hmm, the checking is done on import, so it should only happen once.  In my local scripts, when I build apex without cuda_ext or cpp_ext, the messages are only printed once at the beginning.  Even if I create a loop that calls `import apex` many times, I only see the warning messages printed once.  What does your use case look like?

**Comment by gcp** [comment]
Roughly like this (heavily cut).

I think the errors happen whenever the model switches from eval() to train() and back.

```
#!/usr/bin/env python3

import sys
import os.path
import struct
import random
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import torch.onnx
from apex import amp
from torch.utils.data import Dataset, DataLoader
from torch.autograd import Variable
from timeit import default_timer as timer

....

def main(argv):
    amp_handle = amp.init()
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    net = Net()
    net.load_state_dict(torch.load("net.state"))

    net = torch.nn.DataParallel(net).cuda()

    while failures < 3:  # loop over the dataset multiple times
        net.train()
        for i, data in enumerate(train_loader):
            # zero the parameter gradients
            optimizer.zero_grad()

            with amp_handle.scale_loss(loss, optimizer) as scaled_loss:
                scaled_loss.backward()
            optimizer.step()          

        # Epoch done - Test now
        net.eval()
     
        with torch.no_grad():
            for data in test_loader:

        scheduler.step(loss)       

if __name__ == '__main__':
    main(sys.argv[1:])
```

**Comment by mcarilli** [comment]
I pushed an attempt to fix this https://github.com/NVIDIA/apex/commit/4212b3e943ea8d1b4c0c2749f4b7753de39a6d3a although it still confuses me and I can't reproduce it locally.  Can you re-pull and check if the problem persists?

**Comment by mangotree3** [comment]
In case you have the same problem as me, please see my answer to the problem here: https://github.com/NVIDIA/apex/issues/86#issuecomment-458378045

**Comment by gcp** [comment]
I commented them out locally already.

Thanks for the suggestion but I don't have any intention of changing the OS or switching to containers just to work around this bug.

**Comment by mcarilli** [comment]
Possibly fixed by https://github.com/NVIDIA/apex/commit/48299b0d3c0a56cd7d0b9fed6854d78e81df0cc7.  It was tricky to think of something that wouldn't break existing import statements or suppress errors when they're actually helpful (and the fixes certainly feel hacky) but it works on my machine...let me know if it also works for you.  I have a real love-hate relationship with Python (and i much prefer c++ and cuda) so i'm still open to better ideas.

**Comment by gcp** [comment]
It looks like github master is completely broken on Windows (fails in a LoadLibrary call), I'll try to investigate more later.

**Comment by gcp** [comment]
Fixed in master + PyTorch 1.1.
